walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,102 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  	/** General algorithm (first attempt, could be better optimized, generally runs in linear
+     * time except for calls to Collections.disjoint() in the getAllRelevantEvents() method): 
+     * 1. Get a list of all of the events which actually concern the attendees of the request.
+     * 2. Sort the list by start time.
+     * 3. Loop through, and find all available times "between" these events; requires some amount
+     * of shifting around bounds to deal with edge cases regarding nested events.
+     * 4. Return a list of these times, only adding a time if it is longer than the proposed
+     * duration of the MeetingRequest. 
+     */
+  	Collection<Event> relevantEvents = getAllRelevantEvents(events, request);
+    List<TimeRange> busyTimes = getSortedTimeRanges(relevantEvents);
+    return getComplementTimes(busyTimes, request);
   }
+
+  /**
+   * Returns a Collection of only those events which have
+   * attendees specified in the request parameter.
+   */
+  private Collection<Event> getAllRelevantEvents(Collection<Event> events, MeetingRequest request) {
+    Collection<String> attendees = request.getAttendees();
+    Collection<Event> relevantEvents = new HashSet<>();
+    for (Event event : events) {
+      Collection<String> eventAttendees = event.getAttendees();
+      if (!Collections.disjoint(attendees, eventAttendees)) {
+        relevantEvents.add(event);
+      }
+    }
+    return relevantEvents;
+  }
+
+  /**
+  * Returns a sorted (by start time) List<TimeRange> representing the Collection of events
+  * passed as a parameter. 
+  */
+  private List<TimeRange> getSortedTimeRanges(Collection<Event> events) {
+    List<TimeRange> times = new ArrayList<>();
+    for (Event event : events) {
+      times.add(event.getWhen());
+    }
+    Collections.sort(times, TimeRange.ORDER_BY_START);
+    return times;
+  }
```

Comment:
You could consider combining these instead of (potentially) looping through events twice.

Reply:
I changed it so I no longer loop through events twice, but it still requires two separate methods to convert from a set (to prevent duplicate TimeRanges) to a list (for sorting).